workflow/controller/controller.go

```diff
@@ -446,11 +496,38 @@ func (wfc *WorkflowController) processNextItem() bool {
 		return true
 	}
 
+	if wf.Spec.Synchronization != nil {
+		priority, creationTime := getWfPriority(woc.wf)
+		acquired, wfUpdate, msg, err := wfc.concurrencyMgr.TryAcquire(woc.wf, "", priority, creationTime, woc.wf.Spec.Synchronization)
+		if err != nil {
+			log.Warnf("Failed to acquire the lock for '%s' : %v", key, err)
+			woc.markWorkflowFailed(fmt.Sprintf("invalid spec: %s", err.Error()))
+			woc.persistUpdates()
+			wfc.throttler.Remove(key)
+			return true
+		}
+		woc.updated = wfUpdate
+		if !acquired {
+			log.Warnf("Workflow %s processing has been postponed due to concurrency limit. %s", key, msg)
+			woc.persistUpdates()
+			return true
+		}
+	}
+
 	startTime := time.Now()
 	woc.operate()
+
 	wfc.metrics.OperationCompleted(time.Since(startTime).Seconds())
 	if woc.wf.Status.Fulfilled() {
+		// Release all acquired lock for completed workflow
+		if wfc.concurrencyMgr.ReleaseAll(woc.wf) {
+			log.Infof("%s released all acquired locks", wf.Name)
+			woc.updated = true
+			woc.persistUpdates()
```

Comment:
It doesn't make sense to have both of these next to each other. You probably meant
```suggestion
			woc.updated = true 
```
?

Reply:
Both statements need here. There is no workflow persisting after this section.  woc.update has to be true for persisting in the workflow.